rust/lance/src/index/vector/pq.rs

```diff
@@ -131,6 +131,45 @@ impl PQIndex {
         }))
     }
 
+    /// Pre-compute dot product to each sub-centroids.
+    /// Parameters
+    ///  - query: the query vector, with shape (dimension, )
+    ///  - code: the PQ code in one partition.
+    ///
+    fn dot_distances(&self, key: &Float32Array, code: &UInt8Array) -> Result<ArrayRef> {
+        // Distance table: `[f32: num_sub_vectors(row) * num_centroids(column)]`.
+        let mut distance_table: Vec<f32> = vec![];
+
+        let sub_vector_length = self.dimension / self.num_sub_vectors;
+        for i in 0..self.num_sub_vectors {
+            // The sub-vector section of the query vector.
+            let from = key.slice(i * sub_vector_length, sub_vector_length);
+            let subvec_centroids = self.pq.centroids(i);
+            let distances = dot_distance_batch(
+                as_primitive_array::<Float32Type>(&from).values(),
```

Comment:
Lets use the new api `from.as_primitive().values()` here?

Reply:
Turns out `from` is already primitive so changed to `from.values()`.